File: app/src/main/java/com/jakubwyc/mysecretvault/presenter/VaultPresenter.java
package com.jakubwyc.mysecretvault.presenter;

import android.support.design.widget.Snackbar;
import android.util.Log;

import com.jakubwyc.mysecretvault.model.Treasure;
import com.jakubwyc.mysecretvault.repository.Repository;
import com.jakubwyc.mysecretvault.view.VaultView;

import java.util.ArrayList;
import java.util.List;

public class VaultPresenter implements Presenter<VaultView> {

    private static final String TAG = "VaultPresenter";
    private VaultView view;
    private List<Treasure> treasuresToBeRemoved = new ArrayList<>();


    @Override
    public void attachView(VaultView view) {
        this.view = view;
    }

    @Override
    public void detachView() {
        view = null;
    }

    private boolean isViewAttached() {
        return view != null;
    }

    public void onStart() {
        Repository.getInstance().findTreasures()
                .subscribe(treasures -> {
                    if (isViewAttached()) {
                        view.renderTreasures(treasures);
                    }
                });
    }

    public void onNewTreasureClicked(Void aVoid) {
        view.goToEditTreasureScreen(null);
    }

    public void onEditTreasureClicked(final Treasure treasure) {
        view.goToEditTreasureScreen(treasure);
    }

    public void addCandidateToRemove(final Treasure treasureSwiped) {
        treasuresToBeRemoved.add(treasureSwiped);
    }

    public void removeCandidateToRemove(final Treasure treasureSwiped) {
        treasuresToBeRemoved.remove(treasureSwiped);
    }

    public void removeTreasures(int event) {
        Log.d(TAG, "removeTreasures() called with " + "event = [" + event + "]");
        if (!treasuresToBeRemoved.isEmpty()
                && (event == Snackbar.Callback.DISMISS_EVENT_MANUAL
                || event == Snackbar.Callback.DISMISS_EVENT_TIMEOUT
                || event == Snackbar.Callback.DISMISS_EVENT_ACTION
                || event == Snackbar.Callback.DISMISS_EVENT_SWIPE)) {
            Repository.getInstance().removeTreasures(treasuresToBeRemoved)
                    .subscribe($ -> treasuresToBeRemoved.clear());
        }
    }
}

File: app/src/main/java/com/jakubwyc/mysecretvault/vault/VaultPresenter.kt
package com.jakubwyc.mysecretvault.vault

import com.jakubwyc.mysecretvault.Presenter
import com.jakubwyc.mysecretvault.SystemContext
import com.jakubwyc.mysecretvault.model.Treasure
import rx.subscriptions.CompositeSubscription
import timber.log.Timber

class VaultPresenter(private val systemContext: SystemContext) : Presenter<VaultView> {

    private var view: VaultView? = null
    private val treasuresToBeRemoved = mutableListOf<Treasure>()
    private val subscriptions = CompositeSubscription()
    private val treasureRepository
        get() = systemContext.treasureRepository

    override fun attachView(view: VaultView) {
        this.view = view
    }

    override fun detachView() {
        view = null
    }

    fun onStart() {
        subscriptions.add(
                treasureRepository.findAllTreasures()
                        .subscribe { treasures ->
                            view?.renderTreasures(treasures)
                        })
    }

    fun onNewTreasureClicked() {
        view?.goToEditTreasureScreen(Treasure())
    }

    fun onEditTreasureClicked(treasure: Treasure) {
        view?.goToEditTreasureScreen(treasure)
    }

    fun addCandidateToRemove(treasureSwiped: Treasure) {
        treasuresToBeRemoved.add(treasureSwiped)
    }

    fun removeCandidateToRemove(treasureSwiped: Treasure) {
        treasuresToBeRemoved.remove(treasureSwiped)
    }

    fun removeTreasures() {
        Timber.d("Treasures to be removed: $treasuresToBeRemoved")

        if (treasuresToBeRemoved.isNotEmpty())
            subscriptions.add(
                    treasureRepository.removeAllTreasures(treasuresToBeRemoved.toList())
                            .subscribe { treasuresToBeRemoved.clear() })
    }
}
